File: java/program/code/LinkedList.java
/*
 * NAME: LinkedList
 * AUTHOR: Tanaka Chitete
 * PURPOSE: Implement a generic Linked List
 * CREATION: 31/08/2020
 * LAST MODIFICATION: 09/11/2020
 */

import java.io.Serializable;
import java.util.*;

public class LinkedList<T> extends AbstractSequentialList<T> implements Serializable {
    // NESTED CLASSES

    /*
     * NAME: Node
     * CREATOR: Tanaka Chitete
     * PURPOSE: Implement nodes for LinkedList
     * CREATION: 31/08/2020
     * LAST MODIFICATION: 03/11/2020
     */

    private class Node<A> implements Serializable {
        // PRIVATE CLASS FIELDS
    
        private A element;
        private Node<A> previousNode;
        private Node<A> nextNode;
    
        // CONSTRUCTORS
    
        /*
         * ALTERNATE CONSTRUCTOR
         * IMPORT(S): inElement (Object)
         * EXPORT(S): Address of new Node
         * PURPOSE: Create new Node in alternate state
         * CREATION: 31/08/2020
         * LAST MODIFICATION: 03/11/2020
         */
    
        private Node(A inElement) {
            element = inElement;
            previousNode = null;
            nextNode = null;
        }
    }

    // PRIVATE CLASS FIELDS

    private Node<T> firstNode;
    private Node<T> lastNode;
    private int count;

    // CONSTRUCTORS

    /*
     * DEFAULT CONSTRUCTOR
     * IMPORT(S): NONE
     * EXPORT(S): Address of new LinkedList
     * PURPOSE: Create new LinkedList in default state
     * CREATION: 31/08/2020
     * LAST MODIFICATION: 3/11/2020
     */

    public LinkedList() {
        firstNode = null;
        lastNode = null;
        count = 0;
    }

    // SETTERS (MUTATORS)

    public void insertFirst(T newFirstNodeElement) {
        Node<T> newFirstNode = new Node<T>(newFirstNodeElement);

        if (newFirstNodeElement == null) {
            throw new IllegalArgumentException("Cannot call insertFirst with a null element");
        }
        else {
            if (isEmpty()) {
                firstNode = newFirstNode;
                lastNode = newFirstNode;
            }
            else {
                newFirstNode.nextNode = firstNode;
                firstNode = newFirstNode;
            }
            count++;
        }
    }

    public void insertLast(T newLastNodeElement) {
        Node<T> newLastNode = new Node<T>(newLastNodeElement);

        if (newLastNodeElement == null) {
            throw new IllegalArgumentException("Cannot call insertLast with a null element");
        }
        else {
            if (isEmpty()) {
                lastNode = newLastNode;
                firstNode = newLastNode;
            }
            else {
                lastNode.nextNode = newLastNode;
                lastNode = newLastNode;
            }
            count++;
        }
    }

    // GETTERS (ACCESSORS)

    public T peekFirst() {
        T peekedFirstNodeElement;

        if (isEmpty()) {
            throw new IllegalStateException("Cannot peek at the first node element of an empty " + 
                "linked list");
        }
        else {
            peekedFirstNodeElement = firstNode.element;
        }

        return peekedFirstNodeElement;
    }

    public T peekLast() {
        T peekedLastNodeElement;

        if (isEmpty()) {
            throw new IllegalStateException("Cannot peek at the last node element of an empty " + 
                "linked list");
        }
        else {
            peekedLastNodeElement = lastNode.element;
        }

        return peekedLastNodeElement;
    }

    public T removeFirst() {
        T removedFirstNodeElement;

        if (isEmpty()) {
            throw new IllegalStateException("Cannot remove the first node element of an empty " + 
                "linked list");
        }
        else {
            removedFirstNodeElement = firstNode.element;
            firstNode = firstNode.nextNode;
            count--;
        }

        return removedFirstNodeElement;
    }

    public T removeLast() {
        T removedLastNodeElement;

        if (isEmpty()) {
            throw new IllegalStateException("Cannot remove the last node element of an empty " + 
                "linked list");
        }
        else {
            if (firstNode.nextNode == null) {
                removedLastNodeElement = firstNode.element;
                firstNode = null;
            }
            else {
                removedLastNodeElement = lastNode.element;
                lastNode = lastNode.previousNode;
                lastNode.nextNode = null;
            }
            count--;
        }

        return removedLastNodeElement;
    }

    // OPERATORS

    public String toString() {
        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder.append("[");
        Iterator<T> linkedListIterator = iterator();
        int remainingElements = size();
        while (remainingElements > 0) {
            T currentElement = linkedListIterator.next();
            remainingElements--;
            if (remainingElements == 0) {
                stringBuilder.append(currentElement.toString());
            }
            else {
                stringBuilder.append(currentElement.toString() + ", ");
            }
        }
        stringBuilder.append("]");

        return stringBuilder.toString();
    }

    public boolean isEmpty() {
        return size() == 0;
    }

    @Override
    public int size() {
        return count;
    }   
    
    // ITERATOR

    @Override 
    public ListIterator<T> listIterator(int index) {
        throw new UnsupportedOperationException();
    }
    
    public Iterator<T> iterator() {
        // Hooks iterator to this LinkedList object
        return new LinkedListIterator(this);
    }

    private class LinkedListIterator implements Iterator<T> {
        private Node<T> iterNext;

        public LinkedListIterator(LinkedList<T> inDSALinkedList) {
            iterNext = inDSALinkedList.firstNode;
        }

        @Override
        public boolean hasNext() {
            boolean hasNext;

            if (iterNext == null) {
                hasNext = false;
            }
            else {
                hasNext = true;
            }

            return hasNext;
        }

        @Override
        public T next() {
            T element;

            if (iterNext == null) {
                element = null;
            }
            else {
                element = iterNext.element;
                iterNext = iterNext.nextNode;
            }

            return element;
        }

        public void remove() {
            throw new UnsupportedOperationException();
        }
    }
}

File: python/linked_list.py
"""
NAME: LinkedList
AUTHOR: Tanaka Chitete
PURPOSE: Implement Linked List
CREATION: 27/02/2021
LAST MODIFICATION: 02/03/2021
"""

class LinkedList:
    # NESTED CLASSES

    """
    NAME: LinkedListNode
    AUTHOR: Tanaka Chitete
    PURPOSE: Implement node for Linked List
    CREATION: 27/02/2021 
    LAST MODIFICATION: 02/03/2021
    """

    class LinkedListNode:
        """
        CONSTRUCTOR
        IMPORT(S): element (Object)
        EXPORT(S): Address of new LinkedListNode
        PURPOSE: Make new LinkedListNode
        CREATION: 27/02/2021
        LAST MODIFICATION: 02/03/2021
        """

        def __init__(self, element):
            self.element = element
            self.previous = None
            self.next = None


    """
    NAME: Iterator
    AUTHOR: Tanaka Chitete
    PURPOSE: Implement iterator for Linked List
    CREATION: 01/03/2021 
    LAST MODIFICATION: 2/03/2021
    """

    class Iterator:
        """
        CONSTRUCTOR
        IMPORT(S): linkedList (LinkedList)
        EXPORT(S): Address of new Iterator
        PURPOSE: Make new Iterator
        CREATION: 01/03/2021
        LAST MODIFICATION: 02/03/2021
        """

        def __init__(self, linkedList):
            self.next = linkedList.first


        def hasNext(self):
            return not self.next is None


        def __next__(self):
            element = None
            if self.next is not None:
                element = self.next.element
                self.next = self.next.next
            return element


    """
    CONSTRUCTOR
    IMPORT(S): NONE
    EXPORT(S): Address of new LinkedList
    PURPOSE: Make new LinkedList
    CREATION: 28/02/2021
    LAST MODIFICATION: 02/03/2021
    """

    def __init__(self):
        self.first = None
        self.last = None
        self.count = 0


    # SETTERS (MUTATORS)

    def insertFirst(self, newFirstElement):
        newFirst = self.LinkedListNode(newFirstElement)

        if self.isEmpty():
            self.first = newFirst
            self.last = newFirst
        else:
            newFirst.next = self.first
            self.first = newFirst
        self.count += 1


    def insertLast(self, newLastElement):
        newLast = self.LinkedListNode(newLastElement)

        if self.isEmpty():
            self.last = newLast
            self.first = newLast
        else:
            self.last.next = newLast
            self.last = newLast
        self.count += 1


    # GETTERS (ACCESSORS)

    def peekFirst(self):
        if self.isEmpty():
            raise ValueError("Cannot peek at the first node of an empty linked list")
        else:
            return self.first.element


    def peekLast(self):
        if self.isEmpty():
            raise ValueError("Cannot peek at the last node of an empty linked list")
        else:
            return self.last.element


    def removeFirst(self):
        if self.isEmpty():
            raise ValueError("Cannot remove the first node of an empty linked list")
        else:
            removedFirstElement = self.first.element
            self.first = self.first.next
            self.count -= 1
            return removedFirstElement


    def removeLast(self):
        if self.isEmpty():
            raise ValueError("Cannot remove the last node of an empty linked list")
        elif self.first.next == None:
            removedLastElement = self.first.element
            self.first = None
            self.count -= 1
            return removedLastElement
        else:
            removedLastElement = self.last.element
            self.last = self.last.previous
            self.last.next = None
            self.count -= 1
            return removedLastElement


    # OPERATORS

    def __str__(self):
        nodes = iter(self)
        string = "["
        while nodes.hasNext():
            n = next(nodes)
            # string += f"{str(n)}, "
            string += f"{str(n)} -> " 
        # string = string[:-2] # Removes last ", "
        string = string[:-4] # Removes last " -> "
        string += "]"

        return string


    def __len__(self):
        return self.count


    def __iter__(self):
        return self.Iterator(self)


    def isEmpty(self):
        return self.first is None
